core/frontend/src/tools/SelectTool.ts

```diff
@@ -729,4 +781,262 @@ export class SelectionTool extends PrimitiveTool {
+
+    builderDecoration.setSymbology(color, fillAccDecoration, 2);
+    builderOverlay.setSymbology(colorAccOverlay, fillAccDecoration, 2, LinePixels.Code2);
+
+    // Display area shape
+    if (points.length > 2)
+      builderOverlay.addShape(points);
+
+    // Display line string of area shape which may include a dynamic points
+    builderDecoration.addLineString(points);
+    builderOverlay.addLineString(points);
+
+    // display opened/closed visual helper
+    const openedClosedDeco = context.createGraphicBuilder(GraphicType.ViewOverlay);
+    const firstPtView = context.viewport.worldToView(points[0]);
+    firstPtView.z = 0;
+    const isComplete = this.isSnapToShapeOrigin(ev);
+    openedClosedDeco.setSymbology(color, color, 3);
+    const originSnapRadius = context.viewport.pixelsFromInches(IModelApp.locateManager.apertureInches / 2) + 1.5;
+    openedClosedDeco.addArc(Arc3d.createXY(firstPtView, originSnapRadius), isComplete, isComplete);
+    context.addDecorationFromBuilder(openedClosedDeco);
+
+    context.addDecorationFromBuilder(builderDecoration);
+    context.addDecorationFromBuilder(builderOverlay);
+  }
+
+  private getFullShapePoints(ev: BeButtonEvent): Point3d[] {
+    if (undefined === this.targetView || this._polygonPoints.length < 1)
+      return [];
+
+    const points: Point3d[] = this._polygonPoints.slice();
+
+    // Add the new point if we didn't snap to the origin
+    if (ev && !this.isSnapToShapeOrigin(ev)) {
+      points.push(ev.point.clone());
+    }
+
+    // close the shape
+    if (points.length > 2)
+      points.push(points[0].clone());
+
+    return points;
+  }
+
+  /**
+   * Determine if point can be snapped to the origin of the shape.
+   */
+  private isSnapToShapeOrigin(ev: BeButtonEvent): boolean {
+    if (ev.viewport === undefined || this._polygonPoints.length < 3)
+      return false;
+
+    const firstPtView = ev.viewport.worldToView(this._polygonPoints[0]);
+    const snapDistanceView = ev.viewport.pixelsFromInches(InputSource.Touch === ev.inputSource ? IModelApp.locateManager.touchApertureInches : IModelApp.locateManager.apertureInches);
+    return firstPtView.distanceXY(ev.viewPoint) <= snapDistanceView;
+  }
+
+  /**
+   * Finalize the construction of the polygon, select elements within the polygon, and exit the tool.
+   * @returns true on success
+   */
+  private finishPolygon = async (ev: BeButtonEvent): Promise<boolean> => {
+    if (this._polygonPoints.length < 2) // Handle this event only if we have at least two points
+      return false;
+
+    const ecfPts: XYAndZ[] = [];
+    const ecfXYZPts: XYAndZ[] = [];
+    let minHeight = 0;
+    let maxHeight = 0;
+    let first = true;
+
+    for (const pt of this._polygonPoints) {
+      const ecfPt: Point3d = (this.iModel && this.iModel.isGeoLocated) ? this.iModel.spatialToEcef(pt) : pt;
+      ecfPts.push(ecfPt.toJSONXYZ() as XYAndZ);
+      ecfXYZPts.push(ecfPt.toJSONXYZ() as XYAndZ);
+
+      if (this.iModel.isGeoLocated) {
```

Comment:
This is certainly not correct/needed? Why is this geo located specific code here?